drivers/bus/acpi/eval.c

```diff
@@ -1,255 +1,804 @@
+                        return Status;
+
+                    TotalPackageLength += ElementSize;
+                }
+
+                Argument->Type = ACPI_METHOD_ARGUMENT_PACKAGE;
+                Argument->DataLength = TotalPackageLength;
+
+                Ptr = (PACPI_METHOD_ARGUMENT)Ptr->Data;
+            }
+
+            for (i = 0; i < Obj->Package.Count; i++)
+            {
+                Status = EvalConvertEvaluationResults(Ptr, Depth + 1, &Obj->Package.Elements[i]);
+                if (!NT_SUCCESS(Status))
+                    return Status;
+
+                Ptr = ACPI_METHOD_NEXT_ARGUMENT(Ptr);
+            }
+            break;
+        }
+
+        case ACPI_TYPE_LOCAL_REFERENCE:
+        {
+            Status = EvalConvertObjectReference(Ptr, Obj);
+            if (!NT_SUCCESS(Status))
+                return Status;
+            break;
+        }
+
+        default:
+        {
+            DPRINT1("Unsupported element type %lu\n", Obj->Type);
+
+            return STATUS_UNSUCCESSFUL;
+        }
+    }
+
+    return STATUS_SUCCESS;
+}
+
+static
+CODE_SEG("PAGE")
+ULONG
+EvalGetPackageCount(
+    _In_ PACPI_METHOD_ARGUMENT Package,
+    _In_ PACPI_METHOD_ARGUMENT PackageArgument,
+    _In_ ULONG DataLength)
+{
+    ACPI_METHOD_ARGUMENT* Ptr;
+    ULONG TotalLength = 0, TotalCount = 0;
+
+    PAGED_CODE();
+
+    /* Empty package */
+    if (DataLength < ACPI_METHOD_ARGUMENT_LENGTH(0) || Package->Argument == 0)
+        return 0;
+
+    Ptr = PackageArgument;
+    while (TotalLength < DataLength)
+    {
+        TotalLength += ACPI_METHOD_ARGUMENT_LENGTH_FROM_ARGUMENT(Ptr);
+        TotalCount++;
+
+        Ptr = ACPI_METHOD_NEXT_ARGUMENT(Ptr);
+    }
+
+    return TotalCount;
+}
+
+static
+CODE_SEG("PAGE")
+NTSTATUS
+EvalConvertParameterObjects(
+    _Out_ ACPI_OBJECT* Arg,
+    _In_ ULONG Depth,
+    _In_ PACPI_METHOD_ARGUMENT Argument,
+    _In_ PIO_STACK_LOCATION IoStack,
+    _In_ ULONG Offset)
+{
+    PAGED_CODE();
+
+    if (Depth >= ACPI_MAX_PACKAGE_DEPTH)
+    {
+        ASSERT(FALSE);
+        return STATUS_UNSUCCESSFUL;
+    }
+
+    if (Depth > 0)
+    {
+        Offset += ACPI_METHOD_ARGUMENT_LENGTH_FROM_ARGUMENT(Argument);
+
+        if (!AcpiVerifyInBuffer(IoStack, Offset))
+        {
+            DPRINT1("Argument buffer outside of argument bounds\n");
+
+            return STATUS_ACPI_INVALID_ARGTYPE;
+        }
+    }
+
+    switch (Argument->Type)
     {
-        switch (Package->Package.Elements[i].Type)
+        case ACPI_METHOD_ARGUMENT_INTEGER:
         {
-            case ACPI_TYPE_INTEGER:
-                DPRINT("Integer %lu\n", sizeof(ACPI_METHOD_ARGUMENT));
-                ACPI_METHOD_SET_ARGUMENT_INTEGER(Ptr, Package->Package.Elements[i].Integer.Value);
+            Arg->Type = ACPI_TYPE_INTEGER;
+            Arg->Integer.Value = Argument->Argument;
+            break;
+        }
+
+        case ACPI_METHOD_ARGUMENT_STRING:
+        {
+            /* FIXME: Add tests and remove this */
```

Comment:
Is this still under question?

Reply:
It's really difficult to test this path.